### DISCUSSION TITLE
Is there a way to create good but different trials?

### DISCUSSION THREAD

**Discussion by simpsus** [optuna/optuna#5364] [Q&A] [answered]
In the single objective case:
My goal is not to find the global optimum solution, but a number of good (i.e. close to the optimum) but different (i.e. that the sampled parameters vary a lot) from each other.

Ideas:
Can I write a constraint that discards a trial if the sampled parameters are too similar to already sampled ones?
Can I define a objective value which if reached resets the parameter models to start fresh (with a new random starting point)?

Of course I can use the bruteforce sampler but I would love to have an optimization towards the optimum, albeit with a boundary condition that I do not intend to reach it, but have several good but DIFFERENT candidates.

**Comment by justuswolff-audit** [comment]
- Constraint: imho you'd run into issues where you sample and you keep continuing to not accept the sample parameters. I don't think thats the way to do it.
- Possible in my opinion to implement with a callback to just check whether you've reached a certain objective value.
- Maybe look into weights options of the samplers? I don't exactly understand how they work but that might be an option?

However, what is your assumption about your search space? That there are a lot of different good solutions of similar quality and those are located in local optima?

**Comment by simpsus** [comment]
re callback: I get how I would ensure a "good" candidate (via the objective threshold as you say) but I have no idea how I would tackle the "different" part. I can access the study during a callback, so I can access all trials. I have no idea though how I would go about defining the "distance between two trials in terms of the parameter space" ...

My assumptions on the search space are exactly as you say. Lot's of near-optimal local optimal solutions. I further assume the search space is vast and the global optimum exists, but is pointless to search because it's improvement over the local optima is neglible in contrast to the cost of finding it.

**Comment by justuswolff-audit** [comment]
So I've been thinking about it and don't know if optuna is the right tool for that? I mean it is a multi objective optimisation, where one goal is a high obj. value and the other is finding good but different parameters. I thought of some constant distance measure $d(x^1, x^2)$ where $x^i$ is your parameter selection $i$, that you could define yourself like for $n$ parameters of the following types: boolean, real, integer, categorical. These $n$ parameters are then weighed given their scaling. 
However, how do you find good parameters in optuna? This is done based on the existing trials but you don't wan to search where you already searched (but there you have no clue whether the objective value is actually good). That's why I think this might not be the right tool.
What could work is genetic algorithms (in my opinion). If you set the selection, crossover and mutation parameters properly you should find fairly different solutions (given the assumption of the search space is right). This then fixes the constant distance measure problem as well because this is not used in selection anymore. Then after running the GA, you save the obj. value for each each gene and compute the pairwise distance measures. Given that you could then look into the top objective values sorted by ascending mean distance?
I don't know if that made sense but that is my take on it :)

**Comment by simpsus** [comment, marked as answer]
Thank you very much for these thoughts. I agree with everything you say.
If I use the BruteForceSampler I am guaranteed to get a spread of parameters, but then I am not optimizing! optuna is still an amazing tool to have a study, trials, parameter sampling and the execution of the loop ... but I am not using it like intended.
If I use TPE I will be unhappy with the parameter spread because the sampler constantly wants to converge. 
What I would need would be a switch to say the sampler: That point is locally optimal, please widen your search again. No idea if that is possible. Also no idea if TPE would do that automatically if I constraint-discard like 50 trials in a row. There is an [issue from 2020](https://github.com/optuna/optuna/issues/823) with a similar idea.

As you mention GA: I had this idea as well. But I do not want to start again from scratch. Completely offtopic but: Do you know of a python GA framework I could look into?

Again, thank you very much for the discussion!

**Reply by justuswolff-audit** [reply]
You're welcome. I think the most popular genetic algorithm library in Python is PyGAD. However, I never really found it intuitive and couldn't find anything for optimising mixed variable types (bool, numeric etc.). https://pypi.org/project/geneticalgorithm/ seems fairly promising given the documentation for that task (i just don't know if they support categorical). Personally, I think a basic GA is coded fairly quickly with numpy.
